### PR TITLE
Added expanduser to paths

### LINKED ISSUE #12087: [BUG] Abspath does not expand users home directory
### Software versions

Python version      :  3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10) 
IPython version     :  8.1.1
Tornado version     :  6.1
Bokeh version       :  2.4.2
BokehJS static path :  /home/shh/miniconda3/envs/holoviz/lib/python3.8/site-packages/bokeh/server/static
node.js version     :  v17.8.0
npm version         :  8.5.5
Operating system    :  Linux-5.16.15-76051615-generic-x86_64-with-glibc2.10

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

The code should use `expanduser` in combination with `abspath`.

### Observed behavior

When taking `abspath` of a path starting with the home folder (~) it does not expand the user, and will therefore give a wrong path. 

This is a follow-up to https://github.com/holoviz/panel/pull/3270, which is related to this line:
https://github.com/bokeh/bokeh/blob/14ea3e941229c5069232bf29b48a57f4fb44394a/bokeh/command/util.py#L117

An solution is the following: `os.path.abspath(os.path.expanduser(path))`

### Example code

```Python
import os

path = "~/hello_there.mp4"
a_path = os.path.abspath(path)
ae_path = os.path.abspath(os.path.expanduser(path))
path, a_path, ae_path
```

### Stack traceback or browser console output

_No response_

### PR DESCRIPTION
**Pull request by Hoxbro**
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [x] issues: fixes #12087
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

---
I have only added `expanduser` where the path could come from user input and not something like `os.path.abspath(__file__)`.

### CODE REVIEW

File: versioneer.py [hunk 1]
```diff
@@ -296,7 +296,7 @@ def get_root():
     versioneer_py = os.path.join(root, "versioneer.py")
     if not (os.path.exists(setup_py) or os.path.exists(versioneer_py)):
         # allow 'python path/to/setup.py COMMAND'
-        root = os.path.dirname(os.path.realpath(os.path.abspath(sys.argv[0])))
+        root = os.path.dirname(os.path.realpath(os.path.abspath(os.path.expanduser(sys.argv[0]))))
```
Comment (bryevdv) [inline versioneer.py:299]: This file is vendored and should not be edited (it's also part if the build/dev process and nothing normal users will ever see or interact with)

File: bokeh/util/compiler.py [hunk 2]
```diff
@@ -527,7 +528,7 @@ def resolve_modules(to_resolve: Set[str], root: str) -> Dict[str, str]:
         for module in to_resolve:
             if module.startswith(("./", "../")):
                 def mkpath(module: str, ext: str = "") -> str:
-                    return abspath(join(root, *module.split("/")) + ext)
+                    return abspath(expanduser(join(root, *module.split("/")) + ext))
```
Comment (bryevdv) [inline bokeh/util/compiler.py:531]: I'm not sure a change here is warranted but @mattpap will have to comment when he returns

File: examples/app/stocks/download_sample_data.py [hunk 3]
```diff
@@ -26,7 +26,7 @@ def extract_hosted_zip(data_url, save_dir, exclude_term=None):
 def extract_zip(zip_name, exclude_term=None):
     """Extracts a zip file to its containing directory."""
 
-    zip_dir = os.path.dirname(os.path.abspath(zip_name))
+    zip_dir = os.path.dirname(os.path.abspath(os.path.expanduser(zip_name)))
```
Comment (bryevdv) [inline examples/app/stocks/download_sample_data.py:29]: The download path is path is controlled and always relative to `cwd` this is not needed

File: bokeh/sphinxext/bokeh_gallery.py [hunk 4]
```diff
@@ -120,7 +121,7 @@ def config_inited_handler(app, config):
             continue
 
         with open(detail_file_path, "w") as f:
-            source_path = abspath(join(app.srcdir, "..", "..", detail["path"]))
+            source_path = abspath(expanduser(join(app.srcdir, "..", "..", detail["path"])))
```
Comment (bryevdv) [inline bokeh/sphinxext/bokeh_gallery.py:124]: This is an extension for generating docs, the path is controlled and will never contain a `~`

File: bokeh/application/handlers/code_runner.py [hunk 5]
```diff
@@ -181,7 +181,7 @@ def new_module(self) -> ModuleType | None:
 
         module_name = 'bokeh_app_' + make_globally_unique_id().replace('-', '')
         module = ModuleType(module_name)
-        module.__dict__['__file__'] = os.path.abspath(self._path)
+        module.__dict__['__file__'] = os.path.abspath(os.path.expanduser(self._path))
```
Comment (bryevdv) [inline bokeh/application/handlers/code_runner.py:184]: AFAIK this value is only used for diagnostic output, and I also believe that the only way the `self._path` value originally gets set is from values supplied at the command line, which will already be shell expanded (so this is not necessary).

**Comment by bryevdv** [comment]
@Hoxbro I think this PR may have over-rotated a bit. The only places I think is makes sense to add `expanduser` is where a path that may contain a `~` is specified programmatically by users. In cases where paths are supplied on the command line, they they will automatically shell-exanded before they even reach Bokeh. And in cases where path-construction is controlled (i.e. not by users) there will never be any `~`.

**Comment by Hoxbro** [comment]
I tried to reduce the numbers of `expandpath` I added to the codebase. But clearly, I was too loose with them. I will try to come up with examples (related to bokeh) before adding `expandpath`, and if I cannot, I will remove them. This will properly be done by the end of the week. 

An example of why `expandpath` is needed relating to `bokeh` is the following command will fail (even though the file exists): 
``` bash
$ bokeh serve "~/folder with space/app.py"`
ERROR: Path for Bokeh server application does not exist: /home/shh/~/folder with space/app.py
```

**Comment by bryevdv** [comment]
@Hoxbro Out of curiously what is the realistic use case for using quotes on the cmd line like that?  This is the usage I would expect (which works)
```
dev-3.9 ❯ bokeh serve ~/work/bokeh/examples/app/sliders.py
2022-05-24 11:33:27,376 Starting Bokeh server version 2.4.3 (running on Tornado 6.1)
2022-05-24 11:33:27,377 User authentication hooks NOT provided (default user enabled)
2022-05-24 11:33:27,379 Bokeh app running at: http://localhost:5006/sliders
2022-05-24 11:33:27,379 Starting Bokeh server with process id: 66902
```

That said, I this is also the kind of place where adding it seems OK, since the input is coming from users.

**Comment by Hoxbro** [comment]
I'm not entirely if this is what you are asking for, but the quotes are there because of the spaces in the folder name. If they are not there you get:
``` bash
$ bokeh serve ~/folder with space/app.py  
ERROR: Path for Bokeh server application does not exist: /home/shh/folder
```

File: bokeh/util/compiler.py (hunk 2 above)
Comment (mattpap) [reply bokeh/util/compiler.py:531]: It's unnecessary, perhaps even unsafe to do so.

File: bokeh/io/export.py [hunk 6]
```diff
@@ -119,7 +119,7 @@ def export_png(obj: LayoutDOM | Document, *, filename: PathLike | None = None, w
     filename = os.fspath(filename) # XXX: Image.save() doesn't fully support PathLike
     image.save(filename)
 
-    return abspath(filename)
+    return abspath(expanduser(filename))
```
Comment (Hoxbro) [inline bokeh/io/export.py:122]: `Image.save()` does not work with `~`, or more specifically `open` does not work with `~`. 

I can remove the `expanduser` if this is the desired behavior.

``` python
import os

from bokeh.plotting import figure
from bokeh.io import export_png


p = figure(x_range=(0, 100), y_range=(0, 100), toolbar_location=None)
export_png(p, filename=os.path.expanduser("~/plot.png"))  # Works
export_png(p, filename="~/plot.png")  # Raises FileNotFoundError
```

``` python-traceback
Traceback (most recent call last):
  File "/home/shh/folder with space/save.py", line 7, in <module>
    export_png(p, filename="~/plot.png")
  File "/home/shh/Downloads/bokeh/bokeh/io/export.py", line 120, in export_png
    image.save(filename)
  File "/home/shh/miniconda3/envs/tmp_bokeh/lib/python3.8/site-packages/Pillow-9.1.1-py3.8-linux-x86_64.egg/PIL/Image.py", line 2297, in save
    fp = builtins.open(filename, "w+b")
FileNotFoundError: [Errno 2] No such file or directory: '~/plot.png'
```

File: bokeh/io/saving.py [hunk 7]
```diff
@@ -96,7 +96,7 @@ def save(obj: LayoutDOM, filename: PathLike | None = None, resources: ResourcesL
 
     filename, resources, title = _get_save_args(state, filename, resources, title)
     _save_helper(obj, filename, resources, title, template, theme)
-    return abspath(filename)
+    return abspath(expanduser(filename))
```
Comment (Hoxbro) [inline bokeh/io/saving.py:99]: The same thing as `export_png`, `open` inside `_save_helper` does not work with `~`. 

I can remove the `expanduser` if this is the desired behavior.

``` python
import os

from bokeh.plotting import figure, save

p = figure(x_range=(0, 100), y_range=(0, 100), toolbar_location=None)

filename = save(p, filename=os.path.expanduser("~/index.html"))  # Works
filename = save(p, filename="~/index.html")  # Raises FileNotFoundError

```

``` python-traceback
Traceback (most recent call last):
  File "/home/shh/folder with space/save2.py", line 8, in <module>
    filename = save(p, filename="~/index.html")  # Raises FileNotFoundError
  File "/home/shh/Downloads/bokeh/bokeh/io/saving.py", line 98, in save
    _save_helper(obj, filename, resources, title, template, theme)
  File "/home/shh/Downloads/bokeh/bokeh/io/saving.py", line 166, in _save_helper
    with open(filename, mode="w", encoding="utf-8") as f:
FileNotFoundError: [Errno 2] No such file or directory: '~/index.html'
```

**Comment by bryevdv** [comment]
@Hoxbro the current places all seem appropriate. More or less: `expanduser` is appropriate whenever a path might be user-supplied, e.g a path to save, etc. 

I have restarted the one spurious failed test and will merge when green. FYI if you have any interest in making follow-on PRs, one thing we are slowly working towards is using stdlib `pathlib` for path handling throughout the codebase. Any help towards that would be great.

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
I"m not sure why the one stubborn test won't pass. @Hoxbro can you merge/rebase on latest `branch-3.0` to see if that helps at all?

**Pull request merged**